smartsim/_core/mli/message_handler.py

```diff
@@ -0,0 +1,461 @@
+        """
+        Assigns a reply channel to the supplied request.
+        """
+        try:
+            request.replyChannel.reply = reply_channel
+        except Exception as e:
+            raise ValueError("Error building reply channel portion of request.") from e
+
+    @staticmethod
+    def _assign_device(
+        request: request_capnp.Request, device: t.Union["request_capnp.Device", None]
+    ) -> None:
+        """
+        Assigns a device to the supplied request.
+        """
+        try:
+            if device is None:
+                request.device.noDevice = device
+            else:
+                request.device.deviceType = device
+        except Exception as e:
+            raise ValueError("Error building device portion of request.") from e
+
+    @staticmethod
+    def _assign_inputs(
+        request: request_capnp.Request,
+        inputs: t.Union[
+            t.List[data_references_capnp.TensorKey], t.List[tensor_capnp.Tensor]
+        ],
+    ) -> None:
+        """
+        Assigns inputs to the supplied request.
+        """
+        try:
+            if inputs:
+                display_name = inputs[0].schema.node.displayName  # type: ignore
+                input_class_name = display_name.split(":")[-1]
+                if input_class_name == "Tensor":
+                    request.input.inputData = inputs  # type: ignore
+                elif input_class_name == "TensorKey":
+                    request.input.inputKeys = inputs  # type: ignore
+                else:
+                    raise ValueError(
+                        "Invalid input class name. Expected 'Tensor' or 'TensorKey'."
+                    )
+        except Exception as e:
+            raise ValueError("Error building inputs portion of request.") from e
+
+    @staticmethod
+    def _assign_outputs(
+        request: request_capnp.Request,
+        outputs: t.Optional[t.List[data_references_capnp.TensorKey]],
+    ) -> None:
+        """
+        Assigns outputs to the supplied request.
+        """
+        try:
+            if outputs is not None:
+                request.output.outputKeys = outputs
+            else:
+                request.output.outputData = outputs
+
+        except Exception as e:
+            raise ValueError("Error building outputs portion of request.") from e
+
+    @staticmethod
+    def _assign_output_options(
+        request: request_capnp.Request,
+        output_options: t.List[tensor_capnp.OutputTensorDescriptor],
+    ) -> None:
+        """
+        Assigns a list of output tensor descriptors to the supplied request.
+        """
+        try:
+            request.outputOptions = output_options
+        except Exception as e:
+            raise ValueError(
+                "Error building the output options portion of request."
+            ) from e
+
+    @staticmethod
+    def _assign_custom_request_attributes(
+        request: request_capnp.Request,
+        custom_attrs: t.Union[
```

Comment:
I am also concerned about the extensibility/maintainability of this type! Currently I think we will just need to come back and add to it because I need to check the passed in schema name and set the union field depending on that. So it might be no for now, but hopefully I can come up with a better way in the future.